cluster-autoscaler/cloudprovider/gce/mig_info_provider_test.go

```diff
@@ -0,0 +1,519 @@
+}
+
+func (client *mockAutoscalingGceClient) FetchMachineType(_, _ string) (*gce.MachineType, error) {
+	return nil, nil
+}
+
+func (client *mockAutoscalingGceClient) FetchMachineTypes(_ string) ([]*gce.MachineType, error) {
+	return nil, nil
+}
+
+func (client *mockAutoscalingGceClient) FetchAllMigs(zone string) ([]*gce.InstanceGroupManager, error) {
+	return client.fetchMigs(zone)
+}
+
+func (client *mockAutoscalingGceClient) FetchMigTargetSize(migRef GceRef) (int64, error) {
+	return client.fetchMigTargetSize(migRef)
+}
+
+func (client *mockAutoscalingGceClient) FetchMigBasename(migRef GceRef) (string, error) {
+	return client.fetchMigBasename(migRef)
+}
+
+func (client *mockAutoscalingGceClient) FetchMigInstances(_ GceRef) ([]cloudprovider.Instance, error) {
+	return nil, nil
+}
+
+func (client *mockAutoscalingGceClient) FetchMigTemplateName(migRef GceRef) (string, error) {
+	return client.fetchMigTemplateName(migRef)
+}
+
+func (client *mockAutoscalingGceClient) FetchMigTemplate(migRef GceRef, templateName string) (*gce.InstanceTemplate, error) {
+	return client.fetchMigTemplate(migRef, templateName)
+}
+
+func (client *mockAutoscalingGceClient) FetchMigsWithName(_ string, _ *regexp.Regexp) ([]string, error) {
+	return nil, nil
+}
+
+func (client *mockAutoscalingGceClient) FetchZones(_ string) ([]string, error) {
+	return nil, nil
+}
+
+func (client *mockAutoscalingGceClient) FetchAvailableCpuPlatforms() (map[string][]string, error) {
+	return nil, nil
+}
+
+func (client *mockAutoscalingGceClient) ResizeMig(_ GceRef, _ int64) error {
+	return nil
+}
+
+func (client *mockAutoscalingGceClient) DeleteInstances(_ GceRef, _ []GceRef) error {
+	return nil
+}
+
+func (client *mockAutoscalingGceClient) CreateInstances(_ GceRef, _ string, _ int64, _ []string) error {
+	return nil
+}
+
+func TestGetMigTargetSize(t *testing.T) {
+	targetSize := int64(42)
+	instanceGroupManager := &gce.InstanceGroupManager{
+		Zone:       gceRef.Zone,
+		Name:       gceRef.Name,
+		TargetSize: targetSize,
+	}
+
+	testCases := []struct {
+		name               string
+		cache              *GceCache
+		fetchMigs          func(string) ([]*gce.InstanceGroupManager, error)
+		fetchMigTargetSize func(GceRef) (int64, error)
+		expectedTargetSize int64
+		expectedErr        error
+	}{
+		{
+			name: "target size in cache",
+			cache: &GceCache{
+				migs:               map[GceRef]Mig{gceRef: mig},
+				migTargetSizeCache: map[GceRef]int64{gceRef: targetSize},
+			},
+			expectedTargetSize: targetSize,
+		},
+		{
+			name:               "target size from cache fill",
+			cache:              emptyCache(),
+			fetchMigs:          fetchMigsConst([]*gce.InstanceGroupManager{instanceGroupManager}),
+			expectedTargetSize: targetSize,
+		},
+		{
+			name:               "cache fill without mig, fallback success",
+			cache:              emptyCache(),
+			fetchMigs:          fetchMigsConst([]*gce.InstanceGroupManager{}),
+			fetchMigTargetSize: fetchMigTargetSizeConst(targetSize),
```

Comment:
nit: That feels needlessly clever. It feels like simple targetSize and targetSizeError fields in your struct would accomplish the same thing without the need to rout the interface methods through function references, create closures etc.

I don't think this is worth the effort to refactor and I'm fine merging this as-is. The comment is only meant as something to consider on future PRs.